sdk/schemaregistry/azure-schemaregistry-avroencoder/azure/schemaregistry/encoder/avroencoder/aio/_schema_registry_avro_encoder_async.py

```diff
@@ -0,0 +1,269 @@
+            Indicates an issue with encoding data for provided schema.
+        """
+
+        raw_input_schema = schema
+
+        try:
+            schema_fullname = self._avro_encoder.get_schema_fullname(raw_input_schema)
+        except Exception as e:  # pylint:disable=broad-except
+            SchemaParseError(
+                f"Cannot parse schema: {raw_input_schema}", error=e
+            ).raise_with_traceback()
+
+        http_request_kwargs = get_http_request_kwargs(kwargs)
+        schema_id = await self._get_schema_id(schema_fullname, raw_input_schema, **http_request_kwargs)
+        content_type = f"{AVRO_MIME_TYPE}+{schema_id}"
+
+        try:
+            data_bytes = self._avro_encoder.encode(value, raw_input_schema)
+        except Exception as e:  # pylint:disable=broad-except
+            SchemaEncodeError(
+                "Cannot encode value '{}' for schema: {}".format(
+                    value, raw_input_schema
+                ),
+                error=e,
+            ).raise_with_traceback()
+
+        stream = BytesIO()
+
+        stream.write(data_bytes)
+        stream.flush()
+
+        payload = stream.getvalue()
+        stream.close()
+        if message_type:
+            try:
+                return message_type(data=payload, content_type=content_type, **kwargs)
+            except TypeError as e:
+                SchemaEncodeError(
+                    f"""The data model {str(message_type)} does not support MessageCallbackType protocol.
+                        If using an Azure SDK model class, please check the README.md for the full list
+                        of supported Azure SDK models and their corresponding versions."""
+                ).raise_with_traceback()
+
+        return {"data": payload, "content_type": content_type}
+
+    def _convert_preamble_format(self, data, content_type):  # pylint: disable=no-self-use
+        record_format_identifier = b"\0\0\0\0"
+        if data[0:RECORD_FORMAT_IDENTIFIER_LENGTH] == record_format_identifier:
+            schema_id = data[
+                SCHEMA_ID_START_INDEX : (SCHEMA_ID_START_INDEX + SCHEMA_ID_LENGTH)
+            ].decode("utf-8")
+            content_type = f"{AVRO_MIME_TYPE}+{schema_id}"
+            data = data[DATA_START_INDEX:]
+
+        return data, content_type
+
+    async def decode(
+        self,
+        *,
+        message: Optional[MessageType] = None,
+        data: Optional[bytes] = None,
+        content_type: Optional[str] = None,
+        **kwargs,
+    ) -> Dict[str, Any]:
+        """
+        Decode bytes data using schema ID in the content type field. One of the following is required:
+            1) `message`
+            2) `data` and `content_type`
+        Data must follow format of associated Avro RecordSchema:
+        https://avro.apache.org/docs/1.10.0/gettingstartedpython.html#Defining+a+schema
+
+        :param bytes value: The bytes data needs to be decoded.
```

Comment:
is value needed?

Reply:
so here `value` is the name of the parameter, following the same param name as in the serializer. We may want to discuss naming this data instead for the next release.